### PR TITLE
add ability to publish range of ports

### LINKED ISSUE #8899: Add support for publishing ranges of ports
Given the ability to `--expose` ranges of ports from #8167, it would be useful to be able to publish ranges of ports as well. I'd like to be able to do this:

    docker run -d --publish 10000-20000:10000-20000 myimage

### PR DESCRIPTION
**Pull request by brahmaroutu**
Closes #8899
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

**Comment by brahmaroutu** [comment]
np, I am happy to rebase. Please see that this gets merged.

File: nat/nat.go [hunk 1]
```diff
@@ -122,31 +122,49 @@ func ParsePortSpecs(ports []string) (map[Port]struct{}, map[Port][]PortBinding,
 		if containerPort == "" {
 			return nil, nil, fmt.Errorf("No port specified: %s<empty>", rawPort)
 		}
-		if _, err := strconv.ParseUint(containerPort, 10, 16); err != nil {
+
+		var startPort, endPort uint64 = 0, 0
```
Comment (LK4D4) [inline nat/nat.go]: Still don't understand.

File: integration-cli/docker_cli_create_test.go [hunk 2]
```diff
@@ -101,6 +101,94 @@ func TestCreateHostConfig(t *testing.T) {
...
+		t.Fatalf("Expected HostConfig, got none")
+	}
+
+	if len(c.HostConfig.PortBindings) != 4 {
+		t.Fatalf("Expected 4 ports bindings, got %d", len(c.HostConfig.PortBindings))
+	}
+
+	deleteAllContainers()
+
+	logDone("create - portrange")
```
Comment (LK4D4) [inline integration-cli/docker_cli_create_test.go]: What about this?

File: integration-cli/docker_cli_create_test.go [hunk 3]
```diff
@@ -101,6 +101,94 @@ func TestCreateHostConfig(t *testing.T) {
...
+		t.Fatalf("Expected HostConfig, got none")
+	}
+
+	if len(c.HostConfig.PortBindings) != 65535 {
+		t.Fatalf("Expected 65535 ports bindings, got %d", len(c.HostConfig.PortBindings))
+	}
+
+	deleteAllContainers()
+
+	logDone("create - largeportrange")
```
Comment (LK4D4) [inline integration-cli/docker_cli_create_test.go]: and this

File: integration-cli/docker_cli_create_test.go (hunk 2 above)
Comment (brahmaroutu) [inline integration-cli/docker_cli_create_test.go]: I do not understand, I am doing check on port range, is it what you meant?

Comment (LK4D4) [inline integration-cli/docker_cli_create_test.go]: No, it is just about rewording done message. `port range publishing` instead of `portrange`.

Comment (brahmaroutu) [inline integration-cli/docker_cli_create_test.go]: Yikes, I took @jfrazelle 's comment differently, as we had earlier port checking issues with other bug fixes, I really went ahead and implemented checking on entire port range(for loop). I got this fixed now. Sorry about that.

**Comment by jfrazelle** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by md5** [comment]
:fireworks:

**Comment by thaJeztah** [comment]
\o/

**Pull request merged**